ICE: assertion failed `left: '(Projection, AssocConst)'` `right: ' (ty::Opaque, DefKind::OpaqueTy) | (ty::Projection | ty::Inherent, DefKind::AssocTy) | (ty::Weak, DefKind::TyAlias { .. })`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(associated_type_bounds)]
pub fn accept(_: impl Trait<K :literal>) {}

pub trait Trait {
    const K: i32;
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
307c573d57616a9f7500168061ce6fdbb5060a58
```

### Error output

```
error[E0405]: cannot find trait `literal` in this scope
 --> 16B146BFF37904F6EF9A1E30ACB7C3E39641F49B14B5881AE665829AD30E3805.rs:1:32
  |
1 | pub fn accept(_: impl Trait<K :literal>) {}
  |                                ^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `16B146BFF37904F6EF9A1E30ACB7C3E39641F49B14B5881AE665829AD30E3805`
 --> 16B146BFF37904F6EF9A1E30ACB7C3E39641F49B14B5881AE665829AD30E3805.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `16B146BFF37904F6EF9A1E30ACB7C3E39641F49B14B5881AE665829AD30E3805.rs`
```

If the trait exists, it also crashes *without* debug assertions (stable & nightly):

```rs
pub fn accept(_: impl Trait<K: Copy>) {}

pub trait Trait {
    const K: i32;
}
```

<summary><details>full stderr</details>

```
error[E0658]: associated type bounds are unstable
 --> src/lib.rs:1:29
  |
1 | pub fn accept(_: impl Trait<K: Copy>) {}
  |                             ^^^^^^^
  |
  = note: see issue #52662 <https://github.com/rust-lang/rust/issues/52662> for more information

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_trait_selection/src/traits/project.rs:572:36
stack backtrace:
   0:     0x7f9e36fbd071 - std::backtrace_rs::backtrace::libunwind::trace::h782cc21a5acaf6cb
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f9e36fbd071 - std::backtrace_rs::backtrace::trace_unsynchronized::hc579eb24ab204515
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f9e36fbd071 - std::sys_common::backtrace::_print_fmt::h7223525cfdbacda2
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f9e36fbd071 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbd7d55b7108d2ab8
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f9e3701cd9f - core::fmt::rt::Argument::fmt::hb4f4a02b9bd9dd49
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/fmt/rt.rs:138:9
   5:     0x7f9e3701cd9f - core::fmt::write::h6d54cd7c9e155ec5
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f9e36fb0311 - std::io::Write::write_fmt::h6a453a71c692f63b
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/io/mod.rs:1713:15
   7:     0x7f9e36fbce85 - std::sys_common::backtrace::_print::h1cbaa8b42678f928
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f9e36fbce85 - std::sys_common::backtrace::print::h4ddf81241a51b337
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f9e36fbfb47 - std::panicking::default_hook::{{closure}}::hff91f1f484ade5cd
  10:     0x7f9e36fbf934 - std::panicking::default_hook::h21f14afd59f7aef9
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:288:9
  11:     0x7f9e3a14a0cb - <rustc_driver_impl[a4629e012e48ffc1]::install_ice_hook::{closure#0} as core[95ee2d8017598cc7]::ops::function::FnOnce<(&core[95ee2d8017598cc7]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  12:     0x7f9e36fc028d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha1f76bdae3c27ab6
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1999:9
  13:     0x7f9e36fc028d - std::panicking::rust_panic_with_hook::h45f66047b14c555c
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:709:13
  14:     0x7f9e36fbffe1 - std::panicking::begin_panic_handler::{{closure}}::h49d1a88ef0908eb4
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:595:13
  15:     0x7f9e36fbd4a6 - std::sys_common::backtrace::__rust_end_short_backtrace::hccebf9e57f8cc425
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:151:18
  16:     0x7f9e36fbfd72 - rust_begin_unwind
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:593:5
  17:     0x7f9e37019023 - core::panicking::panic_fmt::h54ec9d0e3180a83d
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:67:14
  18:     0x7f9e370190b3 - core::panicking::panic::h65157a6ac7f1357a
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:117:5
  19:     0x7f9e389a4e96 - <rustc_trait_selection[19572d3103264564]::traits::project::AssocTypeNormalizer as rustc_type_ir[d6fcffcb1d74be50]::fold::TypeFolder<rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt>>::fold_ty
  20:     0x7f9e38d8d2a1 - <rustc_trait_selection[19572d3103264564]::traits::project::AssocTypeNormalizer as rustc_type_ir[d6fcffcb1d74be50]::fold::FallibleTypeFolder<rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt>>::try_fold_binder::<rustc_middle[c682e143f9d710c7]::ty::PredicateKind>
  21:     0x7f9e38d8cfba - <rustc_trait_selection[19572d3103264564]::traits::project::AssocTypeNormalizer as rustc_type_ir[d6fcffcb1d74be50]::fold::FallibleTypeFolder<rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt>>::try_fold_predicate
  22:     0x7f9e38d8cf2d - <core[95ee2d8017598cc7]::iter::adapters::map::Map<alloc[dfb0c3e97fe8b89b]::vec::into_iter::IntoIter<rustc_middle[c682e143f9d710c7]::ty::Predicate>, <alloc[dfb0c3e97fe8b89b]::vec::Vec<rustc_middle[c682e143f9d710c7]::ty::Predicate> as rustc_type_ir[d6fcffcb1d74be50]::fold::TypeFoldable<rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt>>::try_fold_with<rustc_trait_selection[19572d3103264564]::traits::project::AssocTypeNormalizer>::{closure#0}> as core[95ee2d8017598cc7]::iter::traits::iterator::Iterator>::try_fold::<alloc[dfb0c3e97fe8b89b]::vec::in_place_drop::InPlaceDrop<rustc_middle[c682e143f9d710c7]::ty::Predicate>, <core[95ee2d8017598cc7]::iter::adapters::GenericShunt<core[95ee2d8017598cc7]::iter::adapters::map::Map<alloc[dfb0c3e97fe8b89b]::vec::into_iter::IntoIter<rustc_middle[c682e143f9d710c7]::ty::Predicate>, <alloc[dfb0c3e97fe8b89b]::vec::Vec<rustc_middle[c682e143f9d710c7]::ty::Predicate> as rustc_type_ir[d6fcffcb1d74be50]::fold::TypeFoldable<rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt>>::try_fold_with<rustc_trait_selection[19572d3103264564]::traits::project::AssocTypeNormalizer>::{closure#0}>, core[95ee2d8017598cc7]::result::Result<core[95ee2d8017598cc7]::convert::Infallible, !>> as core[95ee2d8017598cc7]::iter::traits::iterator::Iterator>::try_fold<alloc[dfb0c3e97fe8b89b]::vec::in_place_drop::InPlaceDrop<rustc_middle[c682e143f9d710c7]::ty::Predicate>, alloc[dfb0c3e97fe8b89b]::vec::in_place_collect::write_in_place_with_drop<rustc_middle[c682e143f9d710c7]::ty::Predicate>::{closure#0}, core[95ee2d8017598cc7]::result::Result<alloc[dfb0c3e97fe8b89b]::vec::in_place_drop::InPlaceDrop<rustc_middle[c682e143f9d710c7]::ty::Predicate>, !>>::{closure#0}, core[95ee2d8017598cc7]::ops::control_flow::ControlFlow<core[95ee2d8017598cc7]::result::Result<alloc[dfb0c3e97fe8b89b]::vec::in_place_drop::InPlaceDrop<rustc_middle[c682e143f9d710c7]::ty::Predicate>, !>, alloc[dfb0c3e97fe8b89b]::vec::in_place_drop::InPlaceDrop<rustc_middle[c682e143f9d710c7]::ty::Predicate>>>
  23:     0x7f9e38a547ad - rustc_trait_selection[19572d3103264564]::traits::normalize_param_env_or_error
  24:     0x7f9e38a5081f - rustc_ty_utils[f866754e59fc5101]::ty::param_env
  25:     0x7f9e381e9204 - rustc_query_impl[6e478c483b869c25]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6e478c483b869c25]::query_impl::param_env::dynamic_query::{closure#2}::{closure#0}, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 8usize]>>
  26:     0x7f9e381e91d0 - <rustc_query_impl[6e478c483b869c25]::query_impl::param_env::dynamic_query::{closure#2} as core[95ee2d8017598cc7]::ops::function::FnOnce<(rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt, rustc_span[c2ff3b0962b898e5]::def_id::DefId)>>::call_once
  27:     0x7f9e382231a5 - rustc_query_system[93b6edfdc601e81a]::query::plumbing::try_execute_query::<rustc_query_impl[6e478c483b869c25]::DynamicConfig<rustc_query_system[93b6edfdc601e81a]::query::caches::DefaultCache<rustc_span[c2ff3b0962b898e5]::def_id::DefId, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[6e478c483b869c25]::plumbing::QueryCtxt, false>
  28:     0x7f9e39af6503 - rustc_query_impl[6e478c483b869c25]::query_impl::param_env::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7f9e38ba75f0 - rustc_middle[c682e143f9d710c7]::query::plumbing::query_get_at::<rustc_query_system[93b6edfdc601e81a]::query::caches::DefaultCache<rustc_span[c2ff3b0962b898e5]::def_id::DefId, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 8usize]>>>
  30:     0x7f9e38b8c6a9 - rustc_hir_analysis[647275e2a478dd3]::check::wfcheck::check_item_fn
  31:     0x7f9e38b88711 - rustc_hir_analysis[647275e2a478dd3]::check::wfcheck::check_well_formed
  32:     0x7f9e381e5c8c - rustc_query_impl[6e478c483b869c25]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6e478c483b869c25]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 0usize]>>
  33:     0x7f9e38a07746 - rustc_query_system[93b6edfdc601e81a]::query::plumbing::try_execute_query::<rustc_query_impl[6e478c483b869c25]::DynamicConfig<rustc_query_system[93b6edfdc601e81a]::query::caches::VecCache<rustc_hir[2d4be2c3819d30ab]::hir_id::OwnerId, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[6e478c483b869c25]::plumbing::QueryCtxt, false>
  34:     0x7f9e38a07494 - rustc_query_impl[6e478c483b869c25]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7f9e39344c4d - <rustc_middle[c682e143f9d710c7]::hir::ModuleItems>::par_items::<rustc_hir_analysis[647275e2a478dd3]::check::wfcheck::check_mod_type_wf::{closure#0}>
  36:     0x7f9e393448aa - rustc_hir_analysis[647275e2a478dd3]::check::wfcheck::check_mod_type_wf
  37:     0x7f9e38d556ec - rustc_query_impl[6e478c483b869c25]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6e478c483b869c25]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 0usize]>>
  38:     0x7f9e38d556ce - <rustc_query_impl[6e478c483b869c25]::query_impl::check_mod_type_wf::dynamic_query::{closure#2} as core[95ee2d8017598cc7]::ops::function::FnOnce<(rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt, rustc_span[c2ff3b0962b898e5]::def_id::LocalDefId)>>::call_once
  39:     0x7f9e388f67ef - rustc_query_system[93b6edfdc601e81a]::query::plumbing::try_execute_query::<rustc_query_impl[6e478c483b869c25]::DynamicConfig<rustc_query_system[93b6edfdc601e81a]::query::caches::VecCache<rustc_span[c2ff3b0962b898e5]::def_id::LocalDefId, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[6e478c483b869c25]::plumbing::QueryCtxt, false>
  40:     0x7f9e39b00c14 - rustc_query_impl[6e478c483b869c25]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7f9e3961fbe7 - rustc_data_structures[8548742d45a93e9b]::sync::par_for_each_in::<&[rustc_hir[2d4be2c3819d30ab]::hir_id::OwnerId], <rustc_middle[c682e143f9d710c7]::hir::map::Map>::par_for_each_module<rustc_hir_analysis[647275e2a478dd3]::check_crate::{closure#5}::{closure#0}::{closure#0}>::{closure#0}>
  42:     0x7f9e3961f9c7 - <rustc_session[b7d9e6f930bc0bce]::session::Session>::track_errors::<rustc_hir_analysis[647275e2a478dd3]::check_crate::{closure#5}, ()>
  43:     0x7f9e3961cc1f - rustc_hir_analysis[647275e2a478dd3]::check_crate
  44:     0x7f9e39617e2d - rustc_interface[c65fde630b50e8]::passes::analysis
  45:     0x7f9e3965d8fa - rustc_query_impl[6e478c483b869c25]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[6e478c483b869c25]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 1usize]>>
  46:     0x7f9e3965d8e9 - <rustc_query_impl[6e478c483b869c25]::query_impl::analysis::dynamic_query::{closure#2} as core[95ee2d8017598cc7]::ops::function::FnOnce<(rustc_middle[c682e143f9d710c7]::ty::context::TyCtxt, ())>>::call_once
  47:     0x7f9e3981afe8 - rustc_query_system[93b6edfdc601e81a]::query::plumbing::try_execute_query::<rustc_query_impl[6e478c483b869c25]::DynamicConfig<rustc_query_system[93b6edfdc601e81a]::query::caches::SingleCache<rustc_middle[c682e143f9d710c7]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[6e478c483b869c25]::plumbing::QueryCtxt, false>
  48:     0x7f9e3981adb9 - rustc_query_impl[6e478c483b869c25]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  49:     0x7f9e393a5392 - <rustc_interface[c65fde630b50e8]::queries::QueryResult<&rustc_middle[c682e143f9d710c7]::ty::context::GlobalCtxt>>::enter::<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}::{closure#2}::{closure#4}>
  50:     0x7f9e393a4475 - <rustc_interface[c65fde630b50e8]::interface::Compiler>::enter::<rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}::{closure#2}, core[95ee2d8017598cc7]::result::Result<core[95ee2d8017598cc7]::option::Option<rustc_interface[c65fde630b50e8]::queries::Linker>, rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>
  51:     0x7f9e393a25c4 - rustc_span[c2ff3b0962b898e5]::set_source_map::<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_interface[c65fde630b50e8]::interface::run_compiler<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  52:     0x7f9e393a21a1 - <scoped_tls[24938f349c8d16f0]::ScopedKey<rustc_span[c2ff3b0962b898e5]::SessionGlobals>>::set::<rustc_interface[c65fde630b50e8]::interface::run_compiler<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}>::{closure#0}, core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>
  53:     0x7f9e393a1796 - std[edecf719d62c0e6b]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[c65fde630b50e8]::util::run_in_thread_pool_with_globals<rustc_interface[c65fde630b50e8]::interface::run_compiler<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}>::{closure#0}, core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>
  54:     0x7f9e399f94c5 - <<std[edecf719d62c0e6b]::thread::Builder>::spawn_unchecked_<rustc_interface[c65fde630b50e8]::util::run_in_thread_pool_with_globals<rustc_interface[c65fde630b50e8]::interface::run_compiler<core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>, rustc_driver_impl[a4629e012e48ffc1]::run_compiler::{closure#1}>::{closure#0}, core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[95ee2d8017598cc7]::result::Result<(), rustc_span[c2ff3b0962b898e5]::ErrorGuaranteed>>::{closure#1} as core[95ee2d8017598cc7]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  55:     0x7f9e36fca7b5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he3e5dbdfabe0b668
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9
  56:     0x7f9e36fca7b5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h246f7c7964633611
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9
  57:     0x7f9e36fca7b5 - std::sys::unix::thread::Thread::new::thread_start::hadf9e3501ff0df23
                               at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys/unix/thread.rs:108:17
  58:     0x7f9e36e9a609 - start_thread
  59:     0x7f9e36dbd133 - clone
  60:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.1 (eb26296b5 2023-08-03) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [param_env] computing normalized predicates of `accept`
#1 [check_well_formed] checking that `accept` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
For more information about this error, try `rustc --explain E0658`.
error: could not compile `playground` (lib) due to previous error
```

</summary>

--- a/compiler/rustc_hir_analysis/messages.ftl
+++ b/compiler/rustc_hir_analysis/messages.ftl
@@ -1,6 +1,9 @@
 hir_analysis_ambiguous_lifetime_bound =
     ambiguous lifetime bound, explicit lifetime bound required
 
+hir_analysis_assoc_bound_on_const = expected associated type, found {$descr}
+    .note = trait bounds not allowed on {$descr}
+
 hir_analysis_assoc_type_binding_not_allowed =
     associated type bindings are not allowed here
     .label = associated type not allowed here

--- a/compiler/rustc_hir_analysis/src/astconv/bounds.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/bounds.rs
@@ -13,7 +13,7 @@ use crate::astconv::{
     AstConv, ConvertedBinding, ConvertedBindingKind, OnlySelfBounds, PredicateFilter,
 };
 use crate::bounds::Bounds;
-use crate::errors::{MultipleRelaxedDefaultBounds, ValueOfAssociatedStructAlreadySpecified};
+use crate::errors;
 
 impl<'tcx> dyn AstConv<'tcx> + '_ {
     /// Sets `implicitly_sized` to true on `Bounds` if necessary
@@ -35,7 +35,7 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
                     if unbound.is_none() {
                         unbound = Some(&ptr.trait_ref);
                     } else {
-                        tcx.sess.emit_err(MultipleRelaxedDefaultBounds { span });
+                        tcx.sess.emit_err(errors::MultipleRelaxedDefaultBounds { span });
                     }
                 }
             }
@@ -326,7 +326,7 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
             dup_bindings
                 .entry(assoc_item.def_id)
                 .and_modify(|prev_span| {
-                    tcx.sess.emit_err(ValueOfAssociatedStructAlreadySpecified {
+                    tcx.sess.emit_err(errors::ValueOfAssociatedStructAlreadySpecified {
                         span: binding.span,
                         prev_span: *prev_span,
                         item_name: binding.item_name,
@@ -488,6 +488,8 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
             }
         }
 
+        let assoc_item_def_id = projection_ty.skip_binder().def_id;
+        let def_kind = tcx.def_kind(assoc_item_def_id);
         match binding.kind {
             ConvertedBindingKind::Equality(..) if return_type_notation => {
                 return Err(self.tcx().sess.emit_err(
@@ -499,11 +501,9 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
                 // the "projection predicate" for:
                 //
                 // `<T as Iterator>::Item = u32`
-                let assoc_item_def_id = projection_ty.skip_binder().def_id;
-                let def_kind = tcx.def_kind(assoc_item_def_id);
                 match (def_kind, term.unpack()) {
-                    (hir::def::DefKind::AssocTy, ty::TermKind::Ty(_))
-                    | (hir::def::DefKind::AssocConst, ty::TermKind::Const(_)) => (),
+                    (DefKind::AssocTy, ty::TermKind::Ty(_))
+                    | (DefKind::AssocConst, ty::TermKind::Const(_)) => (),
                     (_, _) => {
                         let got = if let Some(_) = term.ty() { "type" } else { "constant" };
                         let expected = tcx.def_descr(assoc_item_def_id);
@@ -516,7 +516,7 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
                             format!("{expected} defined here"),
                         );
 
-                        if let hir::def::DefKind::AssocConst = def_kind
+                        if let DefKind::AssocConst = def_kind
                           && let Some(t) = term.ty() && (t.is_enum() || t.references_error())
                           && tcx.features().associated_const_equality {
                             err.span_suggestion(
@@ -528,8 +528,8 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
                         }
                         let reported = err.emit();
                         term = match def_kind {
-                            hir::def::DefKind::AssocTy => Ty::new_error(tcx, reported).into(),
-                            hir::def::DefKind::AssocConst => ty::Const::new_error(
+                            DefKind::AssocTy => Ty::new_error(tcx, reported).into(),
+                            DefKind::AssocConst => ty::Const::new_error(
                                 tcx,
                                 reported,
                                 tcx.type_of(assoc_item_def_id)
@@ -548,6 +548,15 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
                 );
             }
             ConvertedBindingKind::Constraint(ast_bounds) => {
+                match def_kind {
+                    DefKind::AssocTy => {}
+                    _ => {
+                        return Err(tcx.sess.emit_err(errors::AssocBoundOnConst {
+                            span: assoc_ident.span,
+                            descr: tcx.def_descr(assoc_item_def_id),
+                        }));
+                    }
+                }
                 // "Desugar" a constraint like `T: Iterator<Item: Debug>` to
                 //
                 // `<T as Iterator>::Item: Debug`

--- a/compiler/rustc_hir_analysis/src/errors.rs
+++ b/compiler/rustc_hir_analysis/src/errors.rs
@@ -918,3 +918,12 @@ pub struct UnusedAssociatedTypeBounds {
     #[suggestion(code = "")]
     pub span: Span,
 }
+
+#[derive(Diagnostic)]
+#[diag(hir_analysis_assoc_bound_on_const)]
+#[note]
+pub struct AssocBoundOnConst {
+    #[primary_span]
+    pub span: Span,
+    pub descr: &'static str,
+}

--- a/tests/ui/associated-type-bounds/consts.rs
+++ b/tests/ui/associated-type-bounds/consts.rs
@@ -1,0 +1,10 @@
+#![feature(associated_type_bounds)]
+
+pub fn accept(_: impl Trait<K: Copy>) {}
+//~^ ERROR expected associated type, found associated constant
+
+pub trait Trait {
+    const K: i32;
+}
+
+fn main() {}

--- a/tests/ui/associated-type-bounds/consts.stderr
+++ b/tests/ui/associated-type-bounds/consts.stderr
@@ -1,0 +1,10 @@
+error: expected associated type, found associated constant
+  --> $DIR/consts.rs:3:29
+   |
+LL | pub fn accept(_: impl Trait<K: Copy>) {}
+   |                             ^
+   |
+   = note: trait bounds not allowed on associated constant
+
+error: aborting due to previous error
+

